File: 2017/05.js
const read = require('fs').readFileSync
const fname = '05.txt'

const sm = read(fname, {encoding: 'utf8'})
  .split('\n')
  .map(n => parseInt(n))
sm.pop() // get rid of last blank
const lg = sm.concat([])
const size = sm.length

function small() {
  let i = 0
  let count = 0
  while (i < size) {
    count++
    i = i + sm[i]++
  }

  return count
}

function large() {
  let i, n, count
  i = n = count = 0

  while (i < size) {
    count++
    n = lg[i]
    lg[i] = lg[i] + (n >= 3 ? -1 : 1)
    i = n + i
  }

  return count
}

console.log('small ' + small())
console.log('big ' + large())

File: 2017/05.rb
fname = '05.txt'

class Solution
  @sm = []
  @lg = []
  @size = 0

  def self.add n
    @sm << n
    @lg << n
  end

  def self.sm
    @sm
  end

  def self.lg
    @lg
  end

  def self.size
    @size
  end

  def self.calc_size
    @size = sm.length
  end

  def self.small
    i = count = n = 0
    while i < size do
      count += 1
      n = sm[i]
      sm[i] = sm[i] + 1
      i += n
    end
    count
  end

  def self.large
    i = count = n = 0
    while i < size do
      count += 1
      n = lg[i]
      lg[i] += n >= 3 ? -1 : 1
      i += n
    end
    count
  end

end

File.open(fname, 'r') do |f|
  f.each_line do |line|
    Solution.add(line.to_i)
  end
end

Solution.calc_size
p Solution.small
p Solution.large
